Formal2024/Workshops/workshop6.lean

```diff
@@ -0,0 +1,272 @@
+/-
+Copyright (c) 2024 TheLeanTeam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: The Lean Team
+-/
+
+import Library
+
+/-!
+  # Structures, type classes, and class inference
+  References: [TPIL] Theorem Proving in Lean, [MIL] Mathematics in Lean.
+-/
+
+/-!
+  # Structures
+-/
+
+-- Structures are internally the same thing as inductive types with a single constructor.
+-- They are convenient because they support special features, such as named fields and inheritance.
+
+universe u v
+
+-- Here we define an "α point" (a pair of terms of type α) as an inductive type with a single constructor.
+inductive Point' (α : Type u) : Type u
+| mk : α → α → Point' α
+
+-- We can do the same thing using the `structure` keyword.
+@[ext]
+structure Point (α : Type u) : Type u :=
+mk :: (x : α) (y : α)
+deriving Repr -- this line allows Lean to produce text representations in the infoview.
+
+#check Point
+#check @Point.recOn
+#check @Point.mk -- `mk` is the default constructor name
+
+-- Structures support projection onto their fields
+#check @Point.x
+#check @Point.y
+
+-- Use of the `@[ext]` attribute means Lean generates theorems for proving things by extensionality
+#check Point.ext
+
+example (a b : Point ℕ ) (hx : a.x = b.x) (hy : a.y = b.y) : a = b := by
+  ext
+  repeat' assumption
+
+-- We can print all the details of a structure using `#print prefix`.
+#print prefix Point
+
+-- Notice that the inductive type version has fewer supporting functions.
+#print prefix Point'
+
+-- There are various ways to define instances of a structure.
+-- By providing the named fields directly:
+#check { x := 10, y := 20 : Point ℕ }
+#check { y := 20, x := 10 : Point ℕ }
+#check ({x := 10, y := 20} : Point ℕ)
+
+-- Using the anonymous constructor:
+#check (⟨10, 20⟩ : Point ℕ)
+
+-- Using the default constructor:
+#check (Point.mk 10 20)
+
+-- The anonymous constructor and named fields are useful for defining functions on structures.
+
+namespace Point -- any structure creates a namespace
+
+def add : Point ℕ → Point ℕ → Point ℕ :=
+λ a b => ⟨a.x + b.x, a.y + b.y⟩
+
+def smul (n : ℕ) (p : Point ℕ) :=
+  Point.mk (n * p.x) (n * p.y)
+
+end Point
+
+#check @Point.add
+#check @Point.smul
+
+def p := Point.mk 1 2
+def q := Point.mk 3 4
+
+-- Notice the order in which arguments are input to functions here.
+#eval p.add q
+#check p.add q
+
+#eval p.smul 3
```

Comment:
How does this work? My guess would be that for any function `f` defined in the `Point` namespace taking at least one `p : Point` as an argument, then `p.f` is `Point.f` with `p` as its first `Point` argument? If that is the case, should we write something about that (or maybe it's fine to just mention it during the demo).

Reply:
Yes, your guess is right and I had planned to mention it in the demo. My thinking was that the sheets can get too verbose if those kinds of small details are all mentioned in detail, so my hope is that prompting the reader to think about it and me mentioning it in class is enough.

Reply:
Cool! I think it's fine to just mention it.